Codec/Archive/Tar/Unpack.hs

```diff
@@ -83,32 +85,59 @@ unpack baseDir entries = do
-               HardLink     link -> (unpackEntries $! saveLink True fn link links) es'
-               SymbolicLink link -> (unpackEntries $! saveLink False fn link links) es'
-               OtherEntryType 'L' _ _ -> throwIO $ userError "Two subsequent OtherEntryType 'L'"
-               _ -> unpackEntries links es'
-             (Fail err)      -> either throwIO throwIO err
-             Done            -> throwIO $ userError "././@LongLink without a subsequent entry"
-      _ -> unpackEntries links es --ignore other file types
+    unpackEntries :: Exception e
+                  => Maybe LinkTarget
+                  -> Maybe FilePath
+                  -> [(FilePath, FilePath, Bool)]     -- ^ links (path, link, isHardLink)
+                  -> Entries (Either e FileNameError) -- ^ entries
+                  -> IO [(FilePath, FilePath, Bool)]
+    unpackEntries _ _ _     (Fail err)      = either throwIO throwIO err
+    unpackEntries _ _ links Done            = return links
+    unpackEntries mLink mPath links (Next entry es) = do
+      let path = fromMaybe (entryPath entry) mPath
+      case entryContent entry of
+        NormalFile file _
+          | Just _ <- mLink -> throwIO $ userError "Expected SymbolicLink or HardLink after OtherEntryType K"
+          | otherwise -> do
+              extractFile (entryPermissions entry) path file (entryTime entry)
+              unpackEntries Nothing Nothing links es
+        Directory
+          | Just _ <- mLink -> throwIO $ userError "Expected SymbolicLink or HardLink after OtherEntryType K"
+          | otherwise -> do
+              extractDir path (entryTime entry)
+              unpackEntries Nothing Nothing links es
+        HardLink     link -> do
+          let linkTarget = fromMaybe link mLink
+          (unpackEntries Nothing Nothing $! saveLink True path linkTarget links) es
+        SymbolicLink link -> do
+          let linkTarget = fromMaybe link mLink
+          (unpackEntries Nothing Nothing $! saveLink False path linkTarget links) es
+        OtherEntryType 'L' fn _
+          | Just _ <- mPath -> throwIO $ userError "Two subsequent OtherEntryType L"
+          | otherwise -> unpackEntries mLink (Just . Char8.unpack . BS.toStrict $ fn) links es
+        OtherEntryType 'K' link _
+          | Just _ <- mLink -> throwIO $ userError "Two subsequent OtherEntryType K"
+          | otherwise -> unpackEntries (Just . LinkTarget . BS.toStrict $ link) mPath links es
+        OtherEntryType _ file _
+          | Just _ <- mLink -> throwIO $ userError "Unknown entry type following OtherEntryType K"
+          | Just _ <- mPath -> throwIO $ userError "Unknown entry type following OtherEntryType L"
+          | otherwise -> do
+              hPutStr stderr "Warning: Unknown tar typecode, attempting to extract as normal file"
+              extractFile (entryPermissions entry) path file (entryTime entry)
+              unpackEntries Nothing Nothing links es
```

Comment:
Here we try to adhere to the spec and unpack an unknown entry type as a normal file. Pray.